common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java

```diff
@@ -396,6 +403,56 @@ public void applicationRemoved(String appId, boolean cleanupLocalDirs) {
     }
   }
 
+  @Override
+  public void removeShuffleMerge(RemoveShuffleMerge msg) {
+    AppShuffleInfo appShuffleInfo = validateAndGetAppShuffleInfo(msg.appId);
+    if (appShuffleInfo.attemptId != msg.appAttemptId) {
+      throw new IllegalArgumentException(
+          String.format("The attempt id %s in this RemoveShuffleMerge message does not match "
+                  + "with the current attempt id %s stored in shuffle service for application %s",
+              msg.appAttemptId, appShuffleInfo.attemptId, msg.appId));
+    }
+    appShuffleInfo.shuffles.computeIfPresent(msg.shuffleId, (shuffleId, mergePartitionsInfo) -> {
+      boolean deleteCurrent =
+          msg.shuffleMergeId == DELETE_CURRENT_MERGED_SHUFFLE_ID ||
+              msg.shuffleMergeId == mergePartitionsInfo.shuffleMergeId;
+      AppAttemptShuffleMergeId currentAppAttemptShuffleMergeId =
+          new AppAttemptShuffleMergeId(
+              msg.appId, msg.appAttemptId, msg.shuffleId, mergePartitionsInfo.shuffleMergeId);
+      AppAttemptShuffleMergeId appAttemptShuffleMergeId = new AppAttemptShuffleMergeId(
+          msg.appId, msg.appAttemptId, msg.shuffleId, msg.shuffleMergeId);
+      if(deleteCurrent) {
+        // request to clean up shuffle we are currently hosting
+        if (!mergePartitionsInfo.isFinalized()) {
+          submitCleanupTask(() -> {
+            closeAndDeleteOutdatedPartitions(
+                currentAppAttemptShuffleMergeId, mergePartitionsInfo.shuffleMergePartitions);
+            writeAppAttemptShuffleMergeInfoToDB(appAttemptShuffleMergeId);
+          });
+        } else {
+          submitCleanupTask(() -> {
+            deleteMergedFiles(currentAppAttemptShuffleMergeId,
+                mergePartitionsInfo.getReduceIds());
+            writeAppAttemptShuffleMergeInfoToDB(appAttemptShuffleMergeId);
+            mergePartitionsInfo.setReduceIds(new int[0]);
+          });
+        }
+      } else if(msg.shuffleMergeId < mergePartitionsInfo.shuffleMergeId) {
+        throw new RuntimeException(String.format("Asked to remove old shuffle merged data for " +
+                "application %s shuffleId %s shuffleMergeId %s, but current shuffleMergeId %s ",
+            msg.appId, msg.shuffleId, msg.shuffleMergeId, mergePartitionsInfo.shuffleMergeId));
+      } else if (msg.shuffleMergeId > mergePartitionsInfo.shuffleMergeId) {
+        // cleanup request for newer shuffle - remove the outdated data we have.
+        submitCleanupTask(() -> {
+          closeAndDeleteOutdatedPartitions(
+              currentAppAttemptShuffleMergeId, mergePartitionsInfo.shuffleMergePartitions);
+          writeAppAttemptShuffleMergeInfoToDB(appAttemptShuffleMergeId);
```

Comment:
As mentioned above, `writeAppAttemptShuffleMergeInfoToDB` should be done within the critical section.
We expect the this table to be consistent with `appShuffleInfo.shuffles`

Let us actually move the update to db to the bottom - it is going to be required for all the code paths (since we error out if message is outdated).

Reply:
`The update to db` should be done after `closeAndDeleteOutdatedPartitions` or `deleteMergedFiles` method in `mergedShuffleCleaner` thread, or there will be multi-thread concurrency issue.

Reply:
For the current code, if `currentAppAttemptShuffleMergeId` === `appAttemptShuffleMergeId`
* The main thread run `writeAppAttemptShuffleMergeInfoToDB(appAttemptShuffleMergeId)` before.
* The cleaner task run `removeAppShufflePartitionInfoFromDB(appAttemptShuffleMergeId);`  later, and no `appAttemptShuffleMergeId` in DB

Reply:
I am not sure I understand the concern.

`closeAndDeleteOutdatedPartitions` does two things - 
* cleanup finalization details from DB.
* cleanup files from disk.

Both of these can be done lazily.
We are only keeping `appShuffleInfo.shuffles` consistent with the DB, and use that to filter during recovery.

Note - there is a bug in `finalizeShuffleMerge`, as I mentioned earlier [here](https://github.com/apache/spark/pull/37922#discussion_r990753031) (see the last part of my comment) ... so please keep that in mind when analyzing this codepath.

Reply:
For example, for `appAttemptShuffleMergeId_1  = AppAttemptShuffleMergeId(app_1, appAttempt_1, shuffle_1, shuffleMergeId_1)`, the main thread submits a cleaner task with `appAttemptShuffleMergeId_1`, and then `writeAppAttemptShuffleMergeInfoToDB(appAttemptShuffleMergeId_1)`.
The cleaner thread will cleanup `appAttemptShuffleMergeId_1` from DB later, but we expect `appAttemptShuffleMergeId_1` is in the DB to keep consistent with `appShuffleInfo.shuffles`?